### PR TITLE
Revise Schema Registry docs

### LINKED ISSUE #1311: Revise Schema Registry docs

### PR DESCRIPTION
**Pull request by sobychacko**
Resolves #1311

### CODE REVIEW

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 1]
```diff
@@ -2165,26 +2126,67 @@ public SchemaRegistryClient schemaRegistryClient(@Value("${spring.cloud.stream.s
   return client;
 }
 ----
-NOTE: The ConfluentSchemaRegistryClient is tested against Confluent platform version 3.2.2.
+NOTE: The ConfluentSchemaRegistryClient is tested against Confluent platform version 4.0.0.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: Did we test it against this version?

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 2]
```diff
@@ -2165,26 +2126,67 @@ public SchemaRegistryClient schemaRegistryClient(@Value("${spring.cloud.stream.s
...
-+
-Default:: `true`
+The `spring-cloud-stream-schema` module contains two types of message converters that can be used for Apache Avro serialization:
+
+* Converters that use the class information of the serialized or deserialized objects or a schema with a location known at startup.
+* Converters that use a schema registry. They locate the schemas at runtime and dynamically register new schemas as domain objects evolve.
+
+=== Converters with Schema Support
+
+To use it, you can simply add it to the application context, optionally specifying one or more `MimeTypes` with which to associate it.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: I think it might be useful to define `AvroSchemaMessageConverter` before the usage description.  We can revert to the original order; perhaps we could merge them both into a single statement.

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 3]
```diff
@@ -2165,26 +2126,67 @@ public SchemaRegistryClient schemaRegistryClient(@Value("${spring.cloud.stream.s
...
+
+[source,java]
+----
+@EnableBinding(Sink.class)
+@SpringBootApplication
+public static class SinkApplication {
+
+  ...
+
+  @Bean
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: Maybe we can add a note about why and when this bean is required. That would be "custom" converter requirement, I think.

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 4]
```diff
@@ -1930,11 +1930,117 @@ See "`<<schema-evolution>>`" for details.
 [[schema-evolution]]
 == Schema Evolution Support
 
-Spring Cloud Stream provides support for schema-based message converters through its `spring-cloud-stream-schema` module.
-Currently, the only serialization format supported out of the box for schema-based message converters is Apache Avro, with more formats to be added in future versions.
+Spring Cloud Stream provides support for schema evolution so that the data can be evolved over time and still work with older or newer producers and consumers and vice versa.
+This section goes through the details of various components involved in schema evolution process.
+
+=== Schema Registry Support
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: Maybe we can drop the tile and simply join-in the previous introduction.

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 5]
```diff
@@ -1943,12 +2049,15 @@ The `spring-cloud-stream-schema` module contains two types of message converters
...
 
-To use it, you can simply add it to the application context, optionally specifying one or more `MimeTypes` with which to associate it.
+To use custom converters, you can simply add it to the application context, optionally specifying one or more `MimeTypes` with which to associate it.
 The default `MimeType` is `application/avro`.
 
-The following example shows how to configure a converter in a sink application by registering the Apache Avro `MessageConverter` without a predefined schema:
+If the target type of the conversion is a `GenericRecord`, a schema must be set.
+
+The following example shows how to configure a converter in a sink application by registering the Apache Avro `MessageConverter` without a predefined schema.
+In these examples, note that the mime type value is `avro/bytes`, not the default `application/avro`.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: Probably "this example"? (in singular)

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 6]
```diff
@@ -1930,11 +1930,115 @@ See "`<<schema-evolution>>`" for details.
...
+    String fetch(SchemaReference schemaReference);
+
+    String fetch(Integer id);
+
+}
+----
+
+Spring Cloud Stream provides out-of-the-box implementations for interacting with its own schema server and for interacting with the Confluent Schema Registry.
+
+A client for the Spring Cloud Stream schema registry can be configured byusing the `@EnableSchemaRegistryClient`, as follows:
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: space between "byusing"

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 7]
```diff
@@ -1930,11 +1930,115 @@ See "`<<schema-evolution>>`" for details.
...
+  @EnableBinding(Sink.class)
+  @SpringBootApplication
+  @EnableSchemaRegistryClient
+  public static class AvroSinkApplication {
+    ...
+  }
+----
+
+NOTE: The default converter is optimized to cache not only the schemas from the remote server but also the `parse()` and `toString()` methods, which are quite expensive.
+Because of this, it uses a `DefaultSchemaRegistryClient` that does not caches responses.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: can remove "s" from "caches"

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 8]
```diff
@@ -1930,11 +1930,115 @@ See "`<<schema-evolution>>`" for details.
...
+If you intend to use the client directly on your code, you can request a bean that also caches responses to be created.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: Perhaps rephrase it to:

"If you intend to change the default behavior, you can use the client directly on your code and override it to the desired outcome.

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 9]
```diff
@@ -1930,11 +1930,115 @@ See "`<<schema-evolution>>`" for details.
...
+To do so, add the property `spring.cloud.stream.schemaRegistryClient.cached=true` to your application properties.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: "To do so, you have to add the property ... "

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 10]
```diff
@@ -1930,11 +1930,115 @@ See "`<<schema-evolution>>`" for details.
...
+Default:: `http://localhost:8990/`
+`spring.cloud.stream.schemaRegistryClient.cached`:: Whether the client should cache schema server responses.
+Normally set to `false`, as the caching happens in the message converter.
+Clients using the schema registry client should set this to `true`.
++
+Default:: `true`
+
+=== Avro Schema Registry Client Message Converters
+
+For Spring Boot applications that have a `SchemaRegistryClient` bean registered with the application context, Spring Cloud Stream auto configures an Apache Avro message converter that uses the schema registry client for schema management.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: > For Spring Boot applications that have a `SchemaRegistryClient` bean registered with the application context, Spring Cloud Stream auto configures an Apache Avro message converter that uses the schema registry client for schema management.

to:

> For applications that have a `SchemaRegistryClient` bean registered with the application context, Spring Cloud Stream auto configures an Apache Avro message converter for schema management.

File: spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc [hunk 11]
```diff
@@ -1943,12 +2047,15 @@ The `spring-cloud-stream-schema` module contains two types of message converters
 === Converters with Schema Support
 
 The `AvroSchemaMessageConverter` supports serializing and deserializing messages either by using a predefined schema or by using the schema information available in the class (either reflectively or contained in the `SpecificRecord`).
-If the target type of the conversion is a `GenericRecord`, a schema must be set.
+If the user provides a custom converter (as below), then the default `AvroSchemaMessageConverter` bean will not be created.
```
Comment (sabbyanandan) [inline spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: Much clearer now!

Comment (Buzzardo) [reply spring-cloud-stream-core-docs/src/main/asciidoc/spring-cloud-stream-overview.adoc]: It is much better. To be consistent with the rest of the document, though, I'd change it to: 

If you provide a custom converter, then the default `AvroSchemaMessageConverter` bean is no created. The following example shows a custom converter:

Thanks for putting up with my nitpicking. :)
